halo2_proofs/src/dev/cost_model.rs

```diff
@@ -0,0 +1,364 @@
+        w.write_record(["max_deg", &self.max_deg.to_string()])?;
+        w.write_record(["advice_columns", &self.advice_columns.to_string()])?;
+        w.write_record(["lookups", &self.lookups.to_string()])?;
+        w.write_record(["permutations", &{
+            let mut str = String::new();
+            for p in self.permutations.iter() {
+                str.push_str(&format!(" {}", p.columns));
+            }
+            str
+        }])?;
+        w.write_record(["column_queries", &self.column_queries.to_string()])?;
+        w.write_record(["point_sets", &self.point_sets.to_string()])?;
+        w.write_record([
+            "proof_size",
+            &self.proof_size::<COMM, SCALAR>(comm_scheme).to_string(),
+        ])?;
+        Ok(())
+    }
+}
+
+/// Given a Plonk circuit, this function returns a [ModelCircuit]
+pub fn from_circuit_to_model_circuit<F: Ord + Field + FromUniformBytes<64>, C: Circuit<F>>(
+    k: u32,
+    circuit: &C,
+    instances: Vec<Vec<F>>,
+) -> ModelCircuit {
+    let options = from_circuit_to_cost_model_options(k, circuit, instances);
+    ModelCircuit::from(options)
+}
+
+/// Given a Plonk circuit, this function returns [CostOptions]
+pub fn from_circuit_to_cost_model_options<F: Ord + Field + FromUniformBytes<64>, C: Circuit<F>>(
+    k: u32,
+    circuit: &C,
+    instances: Vec<Vec<F>>,
+) -> CostOptions {
+    let prover = MockProver::run(k, circuit, instances).unwrap();
+    let cs = prover.cs;
+
+    let fixed = {
+        // init the fixed polynomials with no rotations
+        let mut fixed = vec![Poly { rotations: vec![] }; cs.num_fixed_columns()];
+        for (col, rot) in cs.fixed_queries() {
+            fixed[col.index()].rotations.push(rot.0 as isize);
+        }
+        fixed
+    };
+
+    let advice = {
+        // init the advice polynomials with no rotations
+        let mut advice = vec![Poly { rotations: vec![] }; cs.num_advice_columns()];
+        for (col, rot) in cs.advice_queries() {
+            advice[col.index()].rotations.push(rot.0 as isize);
+        }
+        advice
+    };
+
+    let instance = {
+        // init the instance polynomials with no rotations
+        let mut instance = vec![Poly { rotations: vec![] }; cs.num_instance_columns()];
+        for (col, rot) in cs.instance_queries() {
+            instance[col.index()].rotations.push(rot.0 as isize);
+        }
+        instance
+    };
+
+    let lookup = {
+        let mut lookup = vec![];
+        for l in cs.lookups().iter() {
+            lookup.push(Lookup {
+                // this isn't actually used for estimation right now, so ignore it.
+                _columns: 1,
+                input_deg: l.input_expressions().len(),
+                table_deg: l.table_expressions().len(),
+            });
+        }
+        lookup
+    };
+
+    let permutation = vec![Permutation {
+        columns: cs.permutation().get_columns().len(),
+    }];
+
+    let gate_degree = cs.degree();
```

Comment:
This should be `cs.gates().iter().map(Expression::degree).max()`

Reply:
The function degree already gets the max degree of the constrain system. See [here](https://github.com/privacy-scaling-explorations/halo2/blob/main/halo2_proofs/src/plonk/circuit.rs#L2306)